compiler/noirc_evaluator/src/ssa/opt/brillig_entry_points.rs

```diff
@@ -0,0 +1,615 @@
+    for (_, function) in acir_functions {
+        for block_id in function.reachable_blocks() {
+            for instruction_id in function.dfg[block_id].instructions() {
+                let instruction = &function.dfg[*instruction_id];
+                let Instruction::Call { func: func_id, arguments: _ } = instruction else {
+                    continue;
+                };
+
+                let func_value = &function.dfg[*func_id];
+                let Value::Function(func_id) = func_value else { continue };
+
+                let called_function = &functions[func_id];
+                if called_function.runtime().is_acir() {
+                    continue;
+                }
+
+                // We have now found a Brillig entry point.
+                brillig_entry_points.insert(*func_id, BTreeSet::default());
+                build_entry_points_map_recursive(
+                    functions,
+                    *func_id,
+                    *func_id,
+                    &mut brillig_entry_points,
+                    im::HashSet::new(),
+                );
+            }
+        }
+    }
+
+    // If main has been marked as Brillig, it is itself an entry point.
+    // Run the same analysis from above on main.
+    let main_func = &functions[&main_id];
+    if main_func.runtime().is_brillig() {
+        brillig_entry_points.insert(main_id, BTreeSet::default());
+        build_entry_points_map_recursive(
+            functions,
+            main_id,
+            main_id,
+            &mut brillig_entry_points,
+            im::HashSet::new(),
+        );
+    }
+
+    brillig_entry_points
+}
+
+/// Recursively mark any functions called in an entry point
+fn build_entry_points_map_recursive(
+    functions: &BTreeMap<FunctionId, Function>,
+    entry_point: FunctionId,
+    called_function: FunctionId,
+    brillig_entry_points: &mut BTreeMap<FunctionId, BTreeSet<FunctionId>>,
+    mut explored_functions: im::HashSet<FunctionId>,
+) {
+    if explored_functions.insert(called_function).is_some() {
+        return;
+    }
+
+    let inner_calls: HashSet<FunctionId> =
+        called_functions_vec(&functions[&called_function]).into_iter().collect();
+
+    for inner_call in inner_calls {
+        if let Some(inner_calls) = brillig_entry_points.get_mut(&entry_point) {
+            inner_calls.insert(inner_call);
+        }
+
+        build_entry_points_map_recursive(
+            functions,
+            entry_point,
+            inner_call,
+            brillig_entry_points,
+            explored_functions.clone(),
+        );
+    }
+}
+
+/// Builds a mapping from a [`FunctionId`] to the set of [`FunctionId`s][`FunctionId`] of all the brillig entrypoints
+/// from which this function is reachable.
+pub(crate) fn build_inner_call_to_entry_points(
+    brillig_entry_points: &BTreeMap<FunctionId, BTreeSet<FunctionId>>,
+) -> HashMap<FunctionId, Vec<FunctionId>> {
```

Comment:
We likely want to replace this to return a set of function ids rather than a vec.

Reply:
done